pkg/wallet/account/account.go

```diff
@@ -0,0 +1,265 @@
+			Object: &object.Object{
+				Kind:    object.PublicKey,
+				Version: 0x00,
+				Data:    nil,
+			},
+		},
+	}
+}
+
+// NewGenerated generates a new account with a random private key. It destroys the guarded password.
+func NewGenerated(guardedPassword *memguard.LockedBuffer, nickname string) (*Account, error) {
+	version := uint8(AccountLastVersion)
+
+	publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, fmt.Errorf("generating ed25519 key-pair: %w", err)
+	}
+
+	var salt [16]byte
+	_, err = rand.Read(salt[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random salt: %w", err)
+	}
+
+	var nonce [12]byte
+	_, err = rand.Read(nonce[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random nonce: %w", err)
+	}
+
+	aeadCipher, secretKey, err := crypto.NewSecretCipher(guardedPassword.Bytes(), salt[:])
+	guardedPassword.Destroy()
+	if err != nil {
+		return nil, fmt.Errorf("creating secret cipher: %w", err)
+	}
+
+	// secretBuffer is the secret key in clear without the version.
+	secretBuffer := memguard.NewBufferFromBytes(privateKeyBytes.Seed())
+	encryptedSecret := crypto.SealSecret(aeadCipher, nonce[:], secretBuffer)
+	secretKey.Destroy()
+
+	publicKey := types.PublicKey{
+		Object: &object.Object{
+			Kind:    object.PublicKey,
+			Version: types.PublicKeyLastVersion,
+			Data:    publicKeyBytes,
+		},
+	}
+
+	address := types.NewAddressFromPublicKey(&publicKey)
+
+	cipheredData := types.EncryptedPrivateKey{
+		Object: &object.Object{
+			Kind:    object.EncryptedPrivateKey,
+			Version: types.EncryptedPrivateKeyLastVersion,
+			Data:    encryptedSecret,
+		},
+	}
+
+	return New(&version, nickname, *address, salt, nonce, cipheredData, publicKey)
+}
+
+// NewFromPrivateKey creates a new account from a private key. It destroys the guarded password.
+func NewFromPrivateKey(guardedPassword *memguard.LockedBuffer, nickname string, privateKeyText *memguard.LockedBuffer) (*Account, error) {
+	version := uint8(AccountLastVersion)
+
+	var salt [16]byte
+	_, err := rand.Read(salt[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random salt: %w", err)
+	}
+
+	var nonce [12]byte
+	_, err = rand.Read(nonce[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random nonce: %w", err)
+	}
+
+	aeadCipher, secretKey, err := crypto.NewSecretCipher(guardedPassword.Bytes(), salt[:])
+	if err != nil {
+		return nil, fmt.Errorf("creating secret cipher: %w", err)
+	}
+
+	seed, privateKeyVersion, err := base58.CheckDecode(string(privateKeyText.Bytes()[1:])) // omit the first byte because it's 'S' for secret key
+	seedBuffer := memguard.NewBufferFromBytes(seed)
+	if err != nil {
+		seedBuffer.Destroy()
+		return nil, fmt.Errorf("decoding base58 private key: %w", err)
+	}
+
+	privateKey := memguard.NewBufferFromBytes(ed25519.NewKeyFromSeed(seedBuffer.Bytes()))
+	seedBuffer.Destroy()
+
+	encryptedSecret := crypto.SealSecret(aeadCipher, nonce[:], privateKey)
+	secretKey.Destroy()
```

Comment:
I have the feeling that the same code (sealing a private key) is called multiple times. Can you create a hekper to do that only at one place.